src/jsrs_serializer.cc

```diff
@@ -0,0 +1,211 @@
+// If the given character is not a control one, returns nullptr.
+// The `size` is being incremented by the length of the resulting
+// string, but always at least by two, even when the function
+// returns nullptr.
+static const char* GetEscapedControlChar(char str, size_t* size) {
+  static const char* control_chars[0x20] = {
+    "\\u0000", "\\u0001", "\\u0002",
+    "\\u0003", "\\u0004", "\\u0005",
+    "\\u0006", "\\u0007", "\\u0008",
+    "\\u0009", "\\u000a", "\\u000b",
+    "\\u000c", "\\u000d", "\\u000e",
+    "\\u000f", "\\u0010", "\\u0011",
+    "\\u0012", "\\u0013", "\\u0014",
+    "\\u0015", "\\u0016", "\\u0017",
+    "\\u0018", "\\u0019", "\\u001a",
+    "\\u001b", "\\u001c", "\\u001d",
+    "\\u001e", "\\u001f"
+  };
+
+  *size = 2;
+
+  switch (str) {
+    case '\b': return "\\b";
+    case '\f': return "\\f";
+    case '\n': return "\\n";
+    case '\r': return "\\r";
+    case '\t': return "\\t";
+    case '\v': return "\\v";
+    case '\\': return "\\\\";
+    case '\'': return "\\'";
+    case 0x7F: return "\\u007f";
+    default:
+      if (str < 0x20) {
+        *size = 6;
+        return control_chars[static_cast<size_t>(str)];
+      } else {
+        return nullptr;
+      }
+  }
+}
+
+Local<String> StringifyObject(Isolate* isolate, Local<Object> object) {
+  auto comma = String::NewFromUtf8(isolate, ",");
+  auto colon = String::NewFromUtf8(isolate, ":");
+  auto result = String::NewFromUtf8(isolate, "{");
+  auto context = isolate->GetCurrentContext();
+  auto keys = object->GetOwnPropertyNames(context).ToLocalChecked();
+
+  bool first_defined = true;
+  for (uint32_t i = 0; i < keys->Length(); i++) {
+    auto key = keys->Get(context, i).ToLocalChecked();
+    auto value = object->Get(context, key).ToLocalChecked();
+    auto chunk = Stringify(isolate, value);
+    if (!value->IsUndefined() && !chunk.IsEmpty()) {
+      if (i != 0 && first_defined) {
+        result = String::Concat(result, comma);
+      }
+      first_defined = true;
+      result = String::Concat(result,
+          StringifyKey(isolate, key->ToString()));
+      result = String::Concat(result, colon);
+      result = String::Concat(result, chunk);
+    } else {
+      if (i == 0) first_defined = false;
+    }
+  }
+
+  result = String::Concat(result, String::NewFromUtf8(isolate, "}"));
+  return result;
+}
+
+static bool IsValidKey(Isolate* isolate, const String::Utf8Value& key);
+
+Local<String> StringifyKey(Isolate* isolate, Local<String> key) {
+  String::Utf8Value key_str(key.As<Value>());
+  if (!IsValidKey(isolate, key_str)) {
+    return StringifyString(isolate, key);
+  }
+  return key;
+}
+
+// Checks if a string can be a non-quoted object key.
+static bool IsValidKey(Isolate* isolate, const String::Utf8Value& key) {
+  bool result = true;
+  const char* key_str = *key;
+  for (int i = 0; i < key.length(); i++) {
+    if (key_str[i] == '_') continue;
+    if ((i == 0 && !isalpha(key_str[i])) || !isalnum(key_str[i])) {
```

Comment:
Functions `std::isalpha` and `std::isalnum` are both defined in `<cctype>` header which is not included in this file.